File: Samples/Freeform-Tool-Calling/freeform-tool-calling.py
"""
GPT-5 Freeform Tool Calling Demo with Iris Dataset

This script demonstrates GPT-5's new freeform tool calling capabilities using the classic Iris dataset.
Unlike traditional structured tool calling, GPT-5 can now understand natural language instructions
about tools and format its responses accordingly.

Key Features:
- Automatic Iris dataset setup from scikit-learn
- In-memory SQLite database for data analysis
- Python code execution with matplotlib plotting
- Iterative conversation flow with tool results

Requirements:
- Azure OpenAI GPT-5 deployment
- Environment variables: AZURE_OPENAI_ENDPOINT, AZURE_OPENAI_API_KEY, AZURE_OPENAI_DEPLOYMENT
- Python packages: openai, pandas, scikit-learn, matplotlib, python-dotenv
"""

import os, csv, io, sqlite3, textwrap, json
from dataclasses import dataclass
from typing import Dict, Any, List, Optional
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# --- Azure OpenAI client (GPT-5) ---
# You can use API key or Entra ID
# Env vars expected: AZURE_OPENAI_ENDPOINT, AZURE_OPENAI_API_KEY, AZURE_OPENAI_DEPLOYMENT
from openai import AzureOpenAI

# Initialize Azure OpenAI client with GPT-5
# Requires specific API version for freeform tool calling support
client = AzureOpenAI(
    api_key=os.environ["AZURE_OPENAI_API_KEY"],
    api_version="2024-10-01-preview",  # Required API version for GPT-5 freeform tool calling
    azure_endpoint=os.environ["AZURE_OPENAI_ENDPOINT"],
)

MODEL = os.environ["AZURE_OPENAI_DEPLOYMENT"]  # your GPT-5 deployment name

# --- Demo data: iris.csv (create if missing) ---
# Auto-generate the Iris dataset if it doesn't exist locally
# This creates a clean CSV with SQL-friendly column names
IRIS_CSV = "iris.csv"
if not os.path.exists(IRIS_CSV):
    # Write a tiny Iris CSV using sklearn if available; else fail gracefully.
    try:
        import pandas as pd
        from sklearn.datasets import load_iris
        
        # Load the classic Iris dataset from scikit-learn
        iris = load_iris(as_frame=True)
        
        # Rename columns to be SQL-friendly (remove spaces and special characters)
        df = iris.frame.rename(columns={
            "sepal length (cm)": "sepal_length",
            "sepal width (cm)": "sepal_width",
            "petal length (cm)": "petal_length",
            "petal width (cm)": "petal_width",
            "target": "species_id",
        })
        
        # Map numeric species IDs to human-readable names for clarity
        df["species"] = df["species_id"].map({0: "setosa", 1: "versicolor", 2: "virginica"})
        df.drop(columns=["species_id"], inplace=True)
        df.to_csv(IRIS_CSV, index=False)
        print(f"Created {IRIS_CSV} with {len(df)} rows")
    except Exception as e:
        raise SystemExit(
            "Could not auto-create iris.csv. Install scikit-learn & pandas or provide iris.csv.\n"
            f"Error: {e}"
        )

# --- Tool 1: sql_exec_sqlite (raw SQL in, CSV out) ---
# This tool creates an in-memory SQLite database from the CSV file
# and allows GPT-5 to execute SQL queries against it
class SQLiteCSVTool:
    """
    A tool that loads CSV data into an in-memory SQLite database
    and executes SQL queries, returning results as CSV format.
    """
    def __init__(self, csv_path: str):
        # Create in-memory SQLite database for fast querying
        self.conn = sqlite3.connect(":memory:")
        self.conn.row_factory = sqlite3.Row  # Enable column access by name
        self._load_csv(csv_path)

    def _load_csv(self, path: str):
        """Load CSV file into SQLite table named 'iris'"""
        with open(path, "r", newline="", encoding="utf-8") as f:
            reader = csv.DictReader(f)
            cols = [c.strip() for c in reader.fieldnames]
            
            # Create table with all TEXT columns (SQLite will handle type conversion)
            colspec = ", ".join([f'"{c}" TEXT' for c in cols])  # store as TEXT, cast in SQL
            self.conn.execute(f'CREATE TABLE iris ({colspec});')
            
            # Insert all rows into the table
            rows = [tuple(r[c] for c in cols) for r in reader]
            placeholders = ", ".join(["?"] * len(cols))
            self.conn.executemany(f'INSERT INTO iris VALUES ({placeholders});', rows)
            self.conn.commit()

    def exec_sql(self, sql: str) -> str:
        """Execute SQL query and return results as CSV string"""
        cur = self.conn.cursor()
        cur.execute(sql)
        
        # If no rows (e.g., DDL), return empty CSV with a message
        if cur.description is None:
            return "column\n(no result)"
        
        # Convert results to CSV format
        out = io.StringIO()
        writer = csv.writer(out)
        # Write header row with column names
        writer.writerow([d[0] for d in cur.description])
        # Write data rows
        for row in cur.fetchall():
            writer.writerow([row[c[0]] for c in cur.description])
        return out.getvalue()

# Initialize the SQL tool with our Iris dataset
sql_tool = SQLiteCSVTool(IRIS_CSV)

# --- Tool 2: code_exec_python (raw Python in, stdout/PNG note out) ---
# This tool executes Python code and captures output, with matplotlib support
def code_exec_python(code: str, inputs: Optional[Dict[str, Any]] = None) -> str:
    """
    Execute Python code in a sandboxed environment and return stdout.
    Automatically saves matplotlib plots as iris_plot.png.
    
    Args:
        code: Python code string to execute
        inputs: Optional dictionary of inputs available as 'inputs' variable
    
    Returns:
        String containing stdout from code execution
    """
    import contextlib, sys
    import matplotlib
    matplotlib.use("Agg")  # Use non-interactive backend for headless plotting
    import matplotlib.pyplot as plt

    # Capture stdout from code execution
    buf = io.StringIO()
    globs = {"inputs": inputs or {}}  # Make inputs available to executed code
    
    with contextlib.redirect_stdout(buf):
        exec(code, globs, globs)  # Execute user code (demo only - be careful in production!)
        
        # If matplotlib created any figures, save them automatically
        if plt.get_fignums():
            plt.savefig("iris_plot.png", bbox_inches="tight", dpi=150)
            print("\n[Saved figure to iris_plot.png]")
            plt.close("all")
    return buf.getvalue().strip()

# --- GPT-5 Freeform Tool Calling Functions ---
# These functions implement the new freeform tool calling approach
# where GPT-5 uses natural language formatting instead of structured tool schemas

def call_model_with_tools(messages: List[Dict[str, str]]) -> str:
    """
    Call GPT-5 with freeform tool calling enabled.
    
    Unlike traditional tool calling, we don't define structured tool schemas.
    Instead, we rely on GPT-5's ability to understand natural language
    instructions about how to format tool calls using code blocks.
    """
    response = client.chat.completions.create(
        model=MODEL,
        messages=messages,
        # No 'tools' parameter needed - GPT-5 understands tool usage from context
    )
    return response.choices[0].message.content

def parse_and_execute_tools(content: str) -> List[tuple]:
    """
    Parse freeform tool calls from GPT-5 response and execute them.
    
    This function looks for code blocks in GPT-5's response:
    - ```sql blocks for SQL queries
    - ```python blocks for Python code execution
    
    Returns a list of (tool_name, result) tuples.
    """
    results = []
    
    # Look for SQL tool calls in code blocks
    if "```sql" in content:
        # Extract SQL between ```sql and ``` using regex
        import re
        sql_matches = re.findall(r'```sql\n(.*?)\n```', content, re.DOTALL)
        for sql in sql_matches:
            try:
                result = sql_tool.exec_sql(sql.strip())
                results.append(("sql_exec_sqlite", result))
            except Exception as e:
                results.append(("sql_exec_sqlite", f"Error: {e}"))
    
    # Look for Python tool calls in code blocks
    if "```python" in content:
        # Extract Python between ```python and ``` using regex
        import re
        python_matches = re.findall(r'```python\n(.*?)\n```', content, re.DOTALL)
        for code in python_matches:
            try:
                result = code_exec_python(code.strip())
                results.append(("code_exec_python", result))
            except Exception as e:
                results.append(("code_exec_python", f"Error: {e}"))
    
    return results

def run_conversation():
    """
    Main conversation loop demonstrating GPT-5's freeform tool calling.
    
    This function:
    1. Sets up the conversation with system and user prompts
    2. Iteratively calls GPT-5 and executes any tools it requests
    3. Feeds tool results back to continue the conversation
    4. Stops when no more tools are called or max iterations reached
    """
    # System prompt: Explains the freeform tool calling approach to GPT-5
    # Notice how we use natural language to describe tool usage patterns
    system_prompt = """You are a data analyst with access to tools. When you need to execute SQL queries or Python code, format them as code blocks:

For SQL queries, use:
```sql
YOUR_SQL_QUERY_HERE
```

For Python code, use:
```python
YOUR_PYTHON_CODE_HERE
```

You have access to an iris table with columns: sepal_length, sepal_width, petal_length, petal_width, species (all stored as text, cast as needed).
"""

    # User prompt: The specific analysis task we want GPT-5 to complete
    user_prompt = """
1) Write SQL to compute mean of sepal_length, sepal_width, petal_length, petal_width grouped by species.
   Return a tidy CSV with species and the four means (rounded to 2 decimals).
2) Then write Python to read that CSV string (provided as tool output), pretty-print a table,
   and produce a bar chart of mean petal_length by species using Matplotlib.
"""

    # Initialize conversation with system and user messages
    messages = [
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": user_prompt}
    ]

    max_iterations = 5  # Prevent infinite loops
    iteration = 0

    while iteration < max_iterations:
        iteration += 1
        print(f"\n--- Iteration {iteration} ---")
        
        # Get response from GPT-5
        response_content = call_model_with_tools(messages)
        print(f"GPT-5 Response:\n{response_content}")
        
        # Parse and execute any tool calls found in the response
        tool_results = parse_and_execute_tools(response_content)
        
        if not tool_results:
            # No tools were called, we're done with the analysis
            print("\n--- Final Response ---")
            print(response_content)
            break
        
        # Add the assistant's response to the conversation history
        messages.append({"role": "assistant", "content": response_content})
        
        # Add tool results back to the conversation so GPT-5 can continue
        tool_results_text = ""
        for tool_name, output in tool_results:
            tool_results_text += f"\n--- {tool_name} output ---\n{output}\n"
            print(f"\n{tool_name} output:\n{output}")
        
        # Feed tool results back to GPT-5 for continued analysis
        messages.append({"role": "user", "content": f"Tool results:{tool_results_text}\nPlease continue with the analysis."})

if __name__ == "__main__":
    """
    Entry point for the demo.
    
    This script demonstrates:
    - Automatic Iris dataset creation
    - Freeform tool calling with GPT-5
    - SQL analysis of the data
    - Python visualization of results
    
    Make sure you have the required environment variables set:
    - AZURE_OPENAI_ENDPOINT
    - AZURE_OPENAI_API_KEY  
    - AZURE_OPENAI_DEPLOYMENT
    """
    run_conversation()

File: Samples/Freeform-Tool-Calling/TypeScript/freeform-tool-calling.ts
import "dotenv/config";
import { OpenAI } from "openai";
import fs from "fs";
import path from "path";
import { Console } from "console";
import { Writable } from "stream";
import { createContext, runInContext } from "vm";

// Types -----------------------------------------------------------------------

type MessageItem = { role: "system" | "user" | "assistant"; content: string };
type ToolOutputItem = { type: "function_call_output"; call_id: string; output: string };
type ConversationItem = MessageItem | ToolOutputItem;

type IrisRecord = {
  sepal_length: number;
  sepal_width: number;
  petal_length: number;
  petal_width: number;
  species: string;
};

type JavaScriptInputs = {
  csvString?: string;
};

// Environment -----------------------------------------------------------------

const {
  AZURE_OPENAI_ENDPOINT: endpoint,
  AZURE_OPENAI_API_KEY: apiKey,
  AZURE_OPENAI_DEPLOYMENT: model,
} = process.env;

if (!endpoint || !apiKey || !model) {
  throw new Error(
    "AZURE_OPENAI_ENDPOINT, AZURE_OPENAI_API_KEY, and AZURE_OPENAI_DEPLOYMENT environment variables must be set",
  );
}

const deploymentName = model.trim();

const client = new OpenAI({
  apiKey,
  baseURL: `${endpoint}/openai/v1/` // Azure OpenAI v1 is required for latest features
});

// Logging helpers ------------------------------------------------------------

function preview(text: string, max = 200): string {
  const trimmed = text.trim();
  return trimmed.length <= max ? trimmed : `${trimmed.slice(0, max)}…`;
}

// Data helpers ----------------------------------------------------------------

class InMemoryCSVTool {
  private readonly data: IrisRecord[];

  constructor(csvPath: string) {
    this.data = this.loadCSV(csvPath);
  }

  private loadCSV(csvPath: string): IrisRecord[] {
    const csvContent = fs.readFileSync(csvPath, "utf-8");
    const lines = csvContent.trim().split("\n");
    const headers = lines[0].split(",").map((header) => header.trim());

    const records: IrisRecord[] = [];

    for (let i = 1; i < lines.length; i++) {
      const values = lines[i].split(",").map((value) => value.trim());
      const row: Record<string, string | number> = {};

      headers.forEach((header, index) => {
        const rawValue = values[index];
        const numericValue = Number(rawValue);
        row[header] = Number.isNaN(numericValue) ? rawValue : numericValue;
      });

      records.push(row as IrisRecord);
    }

    return records;
  }

  execSQL(sql: string): string {
    const cleanedQuery = sql.trim();
    const lowerQuery = cleanedQuery.toLowerCase();

    if (!lowerQuery.startsWith("select")) {
      return "Error: Only SELECT statements are supported.";
    }

    if (lowerQuery.includes("avg") && lowerQuery.includes("group by") && lowerQuery.includes("species")) {
      return this.computeMeansCSV();
    }

    // Simple fallback to return full dataset as CSV if another SELECT query is executed.
    const headers = Object.keys(this.data[0]);
    const rows = this.data
      .map((row) => headers.map((header) => row[header as keyof IrisRecord]).join(","))
      .join("\n");

    return `${headers.join(",")}\n${rows}`;
  }

  private computeMeansCSV(): string {
    const numericColumns: Array<keyof Omit<IrisRecord, "species">> = [
      "sepal_length",
      "sepal_width",
      "petal_length",
      "petal_width",
    ];

    const aggregates = new Map<string, { count: number; sums: Record<string, number> }>();

    for (const record of this.data) {
      const key = record.species;

      if (!aggregates.has(key)) {
        aggregates.set(key, {
          count: 0,
          sums: Object.fromEntries(numericColumns.map((column) => [column, 0])) as Record<string, number>,
        });
      }

      const entry = aggregates.get(key)!;
      entry.count += 1;

      for (const column of numericColumns) {
        entry.sums[column] += record[column];
      }
    }

    const lines: string[] = [
      [
        "species",
        "sepal_length_mean",
        "sepal_width_mean",
        "petal_length_mean",
        "petal_width_mean",
      ].join(","),
    ];

    for (const [species, { count, sums }] of aggregates.entries()) {
      const formattedMeans = numericColumns.map((column) => (sums[column] / count).toFixed(2));
      lines.push([species, ...formattedMeans].join(","));
    }

    return lines.join("\n");
  }
}

function executeJavaScript(code: string, inputs: JavaScriptInputs): string {
  try {
    let output = "";
    const captureTable = (rows: unknown): void => {
      const buffers: string[] = [];
      const writable = new Writable({
        write(chunk, _encoding, callback) {
          buffers.push(chunk.toString());
          callback();
        },
      });
      const captureConsole = new Console({ stdout: writable, stderr: writable });
      captureConsole.table(rows as any);
      output += buffers.join("");
    };

    const toChartRows = (data: Array<Record<string, unknown>>) =>
      data.map((item, index) => {
        const label = String(item.label ?? item.species ?? `Row ${index + 1}`);
        const rawValue = item.value ?? item.petal_length ?? item.y ?? 0;
        const numericValue = typeof rawValue === "number" ? rawValue : Number(rawValue);
        return { label, value: Number.isFinite(numericValue) ? numericValue : 0 };
      });

    const writeSimpleBarChart = (rows: Array<{ label: string; value: number }>, title: string, filename = "iris_plot.svg") => {
      if (rows.length === 0) {
        throw new Error("Chart data is empty.");
      }

      const width = 480;
      const height = 320;
      const padding = 40;
      const chartWidth = width - padding * 2;
      const chartHeight = height - padding * 2;
      const maxValue = Math.max(...rows.map((row) => row.value), 1);
      const barWidth = chartWidth / rows.length;

      const svgParts = [
        `<?xml version="1.0" encoding="UTF-8"?>`,
        `<svg width="${width}" height="${height}" xmlns="http://www.w3.org/2000/svg">`,
        `<rect width="${width}" height="${height}" fill="white" stroke="#ddd"/>`,
        `<text x="${width / 2}" y="24" text-anchor="middle" font-family="Arial" font-size="16">${title}</text>`,
        `<line x1="${padding}" y1="${height - padding}" x2="${width - padding}" y2="${height - padding}" stroke="#333"/>`,
        `<line x1="${padding}" y1="${padding}" x2="${padding}" y2="${height - padding}" stroke="#333"/>`,
      ];

      rows.forEach((row, index) => {
        const x = padding + index * barWidth + barWidth * 0.1;
        const usableWidth = barWidth * 0.8;
        const scaledHeight = (row.value / maxValue) * chartHeight;
        const y = height - padding - scaledHeight;
        svgParts.push(
          `<rect x="${x.toFixed(2)}" y="${y.toFixed(2)}" width="${usableWidth.toFixed(2)}" height="${scaledHeight.toFixed(2)}" fill="#4E79A7"/>`,
        );
        svgParts.push(
          `<text x="${(x + usableWidth / 2).toFixed(2)}" y="${(y - 5).toFixed(2)}" text-anchor="middle" font-family="Arial" font-size="10">${row.value.toFixed(
            2,
          )}</text>`,
        );
        svgParts.push(
          `<text x="${(x + usableWidth / 2).toFixed(2)}" y="${height - padding / 2}" text-anchor="middle" font-family="Arial" font-size="12">${row.label}</text>`,
        );
      });

      svgParts.push(`</svg>`);

      const chartPath = path.resolve(process.cwd(), filename);
      fs.writeFileSync(chartPath, svgParts.join("\n"), "utf-8");
      output += `[Saved figure to ${chartPath}]\n`;
    };

    const sandbox = createContext({
      inputs,
      csvString: inputs.csvString ?? "",
      console: {
        log: (...args: unknown[]) => {
          output += `${args.map((arg) => (typeof arg === "string" ? arg : JSON.stringify(arg, null, 2))).join(" ")}\n`;
        },
        table: (data: unknown) => {
          captureTable(data);
        },
      },
      JSON,
      Math,
      fs,
      parseCSV: (csv: string) => {
        const trimmed = csv.trim();
        if (!trimmed) return [];

        const [headerLine, ...rowLines] = trimmed.split("\n");
        const headers = headerLine.split(",");
        return rowLines.map((rowLine) => {
          const values = rowLine.split(",");
          return headers.reduce<Record<string, string | number>>((acc, header, index) => {
            const value = values[index];
            const numeric = Number(value);
            acc[header] = Number.isNaN(numeric) ? value : numeric;
            return acc;
          }, {});
        });
      },
      createChart: (data: Array<Record<string, unknown>>, title = "Chart", filename = "iris_plot.svg") => {
        const rows = Array.isArray(data) ? toChartRows(data) : [];
        writeSimpleBarChart(rows, title, filename);
        return rows;
      },
      saveChart: (data: Array<Record<string, unknown>>, title = "Chart", filename = "iris_plot.svg") => {
        const rows = Array.isArray(data) ? toChartRows(data) : [];
        writeSimpleBarChart(rows, title, filename);
        return rows;
      },
    });

    runInContext(code, sandbox);

    return output.trim();
  } catch (error) {
    return `Error: ${error}`;
  }
}

// Tool configuration ----------------------------------------------------------

const TOOLS: Array<Record<string, unknown>> = [
  {
    type: "custom",
    name: "sql_exec_csv",
    description:
      "Executes read-only SQL SELECT queries against the iris dataset and returns tidy CSV results. Use to aggregate data before visualization.",
  },
  {
    type: "custom",
    name: "code_exec_javascript",
    description:
      "Executes Node.js-compatible JavaScript. Use parseCSV(csvString) to read previous SQL output, console.table for inspection, and createChart(data, title) to save an SVG bar chart.",
  },
];

// Conversation setup ---------------------------------------------------------

const IRIS_CSV_PATH = path.resolve(process.cwd(), "iris.csv");

if (!fs.existsSync(IRIS_CSV_PATH)) {
  throw new Error(`${IRIS_CSV_PATH} file not found. Please ensure the iris.csv file exists in the current directory.`);
}

const csvTool = new InMemoryCSVTool(IRIS_CSV_PATH);

const initialMessages: MessageItem[] = [
  {
    role: "system",
    content: `You are a data analyst with access to tools. When you need to execute SQL queries or JavaScript code, format them as code blocks:

For SQL queries, use:
\`\`\`sql
YOUR_SQL_QUERY_HERE
\`\`\`

For JavaScript code, use:
\`\`\`javascript
YOUR_JAVASCRIPT_CODE_HERE
\`\`\`

Important: When writing JavaScript code:
- Write Node.js compatible code (no DOM, no browser APIs like document, canvas, etc.)
- To create charts, use the createChart(data, title) function
- Data should be an array of objects with 'label' and 'value' properties
- Example: createChart([{label: 'setosa', value: 1.46}, {label: 'versicolor', value: 4.26}], 'Petal Length by Species')
- This will automatically save the chart as iris_plot.svg
- The latest SQL tool result is provided as the csvString variable. Always read and parse that (e.g., parseCSV(csvString)) instead of hard-coding data values.


You have access to an iris table with columns: sepal_length, sepal_width, petal_length, petal_width, species (all stored as text, cast as needed).`,
  },
  {
    role: "user",
    content: `1) Write SQL to compute mean of sepal_length, sepal_width, petal_length, petal_width grouped by species.
   Return a tidy CSV with species and the four means (rounded to 2 decimals).
2) Then write JavaScript to read that CSV string (provided as tool output), pretty-print a table,
   and produce a bar chart of mean petal_length by species.`,
  },
];

let lastCSVOutput = "";

async function main(): Promise<void> {
  console.log("🚀 Starting GPT-5 freeform tool calling demo\n");
  console.log(`Using deployment '${deploymentName}' at ${endpoint}`);
  console.log(`Loaded iris dataset from ${IRIS_CSV_PATH}\n`);

  let pending: ConversationItem[] = [...initialMessages];
  let previousResponseId: string | undefined;
  const maxIterations = 5;

  for (let iteration = 1; iteration <= maxIterations && pending.length > 0; iteration += 1) {
    console.log("----------------------------------------");
    console.log(`Iteration ${iteration}: sending ${pending.length} message(s) to GPT-5.`);

    const requestPayload: Parameters<typeof client.responses.create>[0] = {
      model: deploymentName,
      input: pending,
      tools: TOOLS as any,
      text: { format: { type: "text" } },
    };

    if (previousResponseId) {
      requestPayload.previous_response_id = previousResponseId;
    }

    const response = (await client.responses.create(requestPayload)) as any;
    previousResponseId = response?.id ?? previousResponseId;

    pending = [];

    let assistantMessage = "";
    const toolCalls: Array<{ call_id: string; input: string; name: string }> = [];

    for (const item of (response?.output ?? []) as Array<any>) {
      if (item.type === "message") {
        const textFragment = item.content
          ?.map((contentItem: any) => (contentItem.type === "output_text" ? contentItem.text : ""))
          .join("")
          .trim();

        if (textFragment) {
          assistantMessage += `${textFragment}\n`;
        }
      }

      if (item.type === "custom_tool_call") {
        toolCalls.push({ call_id: item.call_id, input: item.input ?? "", name: item.name });
      }
    }

    if (assistantMessage.trim()) {
      const message = assistantMessage.trim();
      console.log(`\nAssistant:\n${message}\n`);
    } else {
      console.log("\nAssistant did not return a direct message this round.\n");
    }

    if (toolCalls.length === 0) {
      console.log("No tool calls requested. Conversation complete.\n");
      break;
    }

    console.log(`Model requested ${toolCalls.length} tool call(s).`);

    for (const toolCall of toolCalls) {
      console.log(`\n[Tool call] ${toolCall.name} (call_id=${toolCall.call_id})`);
      console.log(`Input preview:\n${preview(toolCall.input)}`);

      const result = executeTool(toolCall.name, toolCall.input);
      console.log(`Output:\n${result.trim()}`);

      if (toolCall.call_id.startsWith("auto-")) {
        const userMessage: MessageItem = {
          role: "user",
          content: `[System] Executed code block:\n${result}`,
        };
        pending.push(userMessage);
      } else {
        const toolOutput: ToolOutputItem = {
          type: "function_call_output",
          call_id: toolCall.call_id,
          output: result,
        };
        pending.push(toolOutput);
      }
    }
  }

  if (pending.length > 0) {
    console.log("Reached iteration limit before conversation completed.\n");
  }

  console.log("Demo complete.");
}

function executeTool(name: string, input: string): string {
  switch (name) {
    case "sql_exec_csv": {
      const sql = input.trim();
      const output = csvTool.execSQL(sql);
      if (!output.startsWith("Error")) {
        lastCSVOutput = output;
      }
      return output;
    }
    case "code_exec_javascript": {
      if (!lastCSVOutput) {
        return "Error: No CSV data available. Call sql_exec_csv first.";
      }
      const code = input.trim();
      return executeJavaScript(code, { csvString: lastCSVOutput });
    }
    default:
      return `Error: Unknown tool '${name}'.`;
  }
}

main().catch((error) => {
  console.error("The sample encountered an error:", error);
  process.exit(1);
});
